lib/client/configurator/util.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package configurator
+
+import (
+	"os"
+	"os/exec"
+
+	"github.com/gravitational/trace"
+)
+
+// ensureSymlinks recreates the provided symlinks.
+func ensureSymlinks(symlinks map[string]string) error {
+	for source, destination := range symlinks {
+		if err := ensureSymlink(source, destination); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+// verifySymlinks verifies a set of symlinks.
+func verifySymlinks(symlinks map[string]string) (bool, error) {
+	for source, destination := range symlinks {
+		if ok, err := verifySymlink(source, destination); !ok || err != nil {
+			return false, trace.Wrap(err)
+		}
+	}
+	return true, nil
+}
+
+// ensureSymlinks recreates the specified symlink.
+func ensureSymlink(source, destination string) error {
+	log.Debugf("Symlinking %v -> %v.", source, destination)
+	if err := os.RemoveAll(destination); err != nil {
+		return trace.ConvertSystemError(err)
+	}
+	if err := os.Symlink(source, destination); err != nil {
+		return trace.ConvertSystemError(err)
+	}
+	return nil
+}
+
+// verifySymlink verifies that the symlink exists and points to the expected location.
+func verifySymlink(source, destination string) (bool, error) {
+	fileInfo, err := os.Lstat(destination)
+	err = trace.ConvertSystemError(err)
+	if err != nil && !trace.IsNotFound(err) {
+		return false, trace.Wrap(err)
+	}
+	if trace.IsNotFound(err) {
+		return false, nil
+	}
+	// If the file exists, it must be a symlink and point to the specified location.
+	if fileInfo.Mode()&os.ModeSymlink == 0 {
+		return false, nil
+	}
+	actualSource, err := os.Readlink(destination)
+	if err != nil {
+		return false, trace.ConvertSystemError(err)
+	}
+	if actualSource != source {
+		return false, nil
+	}
+	return true, nil
+}
+
+// runCommandSudo executes the specified command as sudo.
```

Comment:
I don't think we should ever hardcode as sudo, instead ask user to run as `sudo` or any other privileges.

Reply:
@klizhentas So this runs as a part of `tsh login` process and thus if we don't want to invoke sudo here, I see 2 options:

* Ask the user to run the entire `tsh login` as root which I'm pretty sure we don't wanna do.
* Ask the user to run `tsh docker configure` themselves as a root afterwards which will sort of interrupt the login flow.

With invoking the command as a sudo here, I tried to limit the "scope" that runs as a root while still preserving decent UX because they will be auto-prompted for password if needed.

We could make this more explicit by sacrificing a bit on the UX side, for example:

* User runs `tsh login` which fetches certs and configures helm (which doesn't need root).
* Next we see if `tsh` has permissions to symlink to /etc/docker/certs.d.
  * If it does (maybe it's already running as root) - everything is cool, we do the symlinking.
  * If it does not (which likely would be most of the time) - we just output the message at the end asking the user to run `tsh configure docker` as a user that has permissions to /etc/docker/certs.d if they wish to configure registry access. This should only be needed the first time b/c for subsequent logins symlinks will already be there (unless removed).

Would that be better?

Reply:
I'm concerned about `tsh` attempting to elevate it's own privileges without user's approval as a pattern. @knisbet @russjones what do you folks think about it?

Reply:
I like @r0mant's suggestion, tsh can print a message explaining what it needs to do (create symlinks as root) and then ask the user to run the following command as `sudo` to do it.

Reply:
My 2 cents, is I think docker configure should be a separate command tsh command, and if run as non-root it can then prompt indicating it should be run as root and why (or link to a KB article or something on what happens.

The reason I think this:
- I suspect it may be unsafe to expect the user to have root permissions on their host on login. Some orgs may have laptops or machines where TSH is used hardened in a way that prevents sudo. I'm not sure we can predict what sort of policies may exist or that calling sudo will be available.
- I'm not sure the docker login will be a commonly used feature. I suspect the average user of a gravity cluster does not need access to the images. So I don't have an issue with having a separate configuration command that is run once with matching docs.
- I'm of the opinion it's a cleaner UX pattern, I think some of the miss-understandings of our product come from trying to cleverly do extra things that are non-intuitive. I didn't indicate I want to log into the docker registry, but on tsh login this other login happens automatically / behind the scenes. It would be different if this was a common pattern or a commonly understood part of the intentions.
- I think this is a normal approach in other tools, `gcloud compute config-ssh` comes to mind as something that after login I knowingly execute to update my ssh configuration.
- It only needs to be done once, so I don't see it as really interrupting the flow.

Downsides of this approach are I think the standard things the team tries to avoid, which is having additional commands in the cli, how those commands are exposed / seen on non-gravity clusters, that this is something a potential user must discover instead of being "magic", how the configuration gets removed, that a user remembers to run it on a new machine, etc.

Possible alternatives:
- The docker setup could be a flag to tsh login, so it's intention based. I'm not personally a big fan of this, but it would still create a clearer intention from the user that they intend for the login to setup docker and associated requirements.
- I'm not opposed to the proposal of printing in the tsh login how to configure docker at the end of the login if not already configured. My opinion though is that it probably adds noise to the login process for the average user.

Reply:
@benarent I think this has to be reviewed by your team as well. My main concern is as is this PR asks `tsh login` to elevate it's privileges and re-run itself as root. However, alternatives outlined here should be visible to customer, so you should take a look into this. In addition to that, it would be helfpul to support a case when users don't actually want their docker / helm credentials updated. Can you please work with @r0mant to figure out the UX of this part and update this PR with your thoughts on the matter.

Reply:
@klizhentas @russjones @knisbet We have spent some time today brainstorming with @benarent on this and updated the design doc with what we came up with: https://docs.google.com/document/d/1V2bI2oZQC3Y3l42_eOvwJZSrPykN6Oxm0vO4oWDiv78/edit#heading=h.ykw97ikonjby (see **tsh UX** part at the bottom).

The basic flow is like I described above i.e. to ask user to run the "configure" command explicitly in case they don't have permissions.

To support the use-case where they may not want to have their docker/helm configured, instead of going with the "add more flags" approach, we opted for placing these options into the role, similar to "agent/port forwarding" options it has now. This way they would be able to create a role that does not configure docker/helm by default, but still have an option to configure them manually via specific "configure" commands if they want. These options will be encoded into certs, again similar to agent/port forwarding and other stuff.

Reply:
@r0mant @benarent I think role options is not a great place for what is basically user preferences. The option of logging into helm or docker registry is a client-level preference, for example same user may or may not want to login into docker on different machines.

I think a good place for user preferences is default global login preferences section in our configuration file duplicated in `.tsh/profile`